not-script/not-script.c

```diff
@@ -0,0 +1,414 @@
+
+    int dev_file = -1, status;
+    for (int retry_count = 0; retry_count < 5 /* arbitrary */; retry_count++) {
+        if ((status = ioctl(ctx->fd, LOOP_CTL_GET_FREE)) == -1)
+            return -errno;
+        if ((unsigned)snprintf(buf, sizeof buf, "/dev/loop%u", (unsigned)status) >= sizeof buf)
+            abort();
+        dev_file = open(buf, (writable ? O_RDWR : O_RDONLY) |
+                              O_EXCL | O_CLOEXEC | O_NOCTTY | O_NOFOLLOW);
+        if (dev_file > 0) {
+            switch (ioctl(dev_file, LOOP_CONFIGURE, &config)) {
+            case 0:
+                return dev_file;
+            case -1:
+                if (close(dev_file))
+                    abort(); /* cannot happen on Linux */
+                return -1;
+            default:
+                abort();
+            }
+        }
+        if (errno != EBUSY)
+            break;
+    }
+    return -1;
+}
+
+#define DEV_MAPPER "/dev/mapper/"
+#define DEV_MAPPER_SIZE (sizeof DEV_MAPPER - 1)
+
+static void
+check_sysfs_attr(dev_t const rdev,
+                 const char *const sysfs_attr,
+                 const char *const expected,
+                 const char *const devtype)
+{
+    char *sysfs_path;
+
+    if (asprintf(&sysfs_path, "/sys/dev/block/%" PRIu32 ":%" PRIu32 "/%s",
+                 major(rdev), minor(rdev), sysfs_attr) < 0)
+        err(1, "asprintf");
+
+    int const sysfs_fd = open(sysfs_path, O_RDONLY | O_NOCTTY | O_NOFOLLOW);
+    if (sysfs_fd < 0)
+        err(1, "open(%s)", sysfs_path);
+
+    size_t const devname_len = strlen(expected);
+
+    char *const cmp_ptr = malloc(devname_len + 2);
+    if (!cmp_ptr)
+        err(1, "malloc()");
+
+    ssize_t const read_res = read(sysfs_fd, cmp_ptr, devname_len + 2);
+    if (read_res < 0)
+        err(1, "read(%s)", sysfs_path);
+
+    if (((size_t)read_res != devname_len + 1) ||
+        (memcmp(expected, cmp_ptr, devname_len) != 0) ||
+        (cmp_ptr[devname_len] != '\n'))
+    {
+        errx(1, "Opened the wrong %s device!", devtype);
+    }
+
+    free(cmp_ptr);
+    if (close(sysfs_fd))
+        err(1, "close()");
+}
+
+static void
+process_blk_dev(int fd, const char *path, bool writable, dev_t *dev,
+                uint64_t *diskseq, bool permissive)
+{
+    struct stat statbuf;
+
+    if (fstat(fd, &statbuf))
+        err(1, "fstat");
+
+    if (S_ISBLK(statbuf.st_mode)) {
+        /* block device */
+        const char *sysfs_attr = "dm/name";
+        const char *devtype = "device-mapper";
+        const char *expected;
+
+        if (permissive) {
+            char *qubesdb_path;
+            const char *extra_path;
+
+            if (strncmp(path, "/dev/", 5))
+                errx(1, "Bad path %s", path);
+
+            extra_path = path + 5;
+
+            qdb_handle_t h = qdb_open(NULL);
+
+            if (!h)
+                err(1, "qdb_open(NULL)");
+
+            if (asprintf(&qubesdb_path, "/qubes-block-devices/%s/desc", extra_path) < 0)
```

Comment:
I doubt checking against qubesdb is helpful at all. For accidental race to happen, user action is very likely to be involved (and user wouldn't do it between clicking attach and the actual attach). And even if it would, then udev trigger is very likely to update the qubesdb entry before this check happens (the window when `/dev/loop*` points at a different thing than qubesdb says is _very_ small).
As for malicious race (backend trying to make user attach a different thing), this won't help at all, as backend process (especially capable of changing loop devices) can simply replace file content.

What could be useful, is to generate names that are somehow bound to the content (like `loop-XXXX` where `XXXX` is some short checksum of the target path, even crc32 would do), but that's out of scope of this PR.

Reply:
I will submit a PR to use `X-old-name`, where `X` is the device diskseq.  The not-script can check for that.  At some point I want to rewrite the udev trigger script in something other than shell, perhaps C.